Refactor getting started pages

nojnhuh opened the pull request:
- remove references to VMs
- remove content from Getting Started index page
- Manual demo
  - rename bookstore service backend to bookstore-v1
  - add SMI policies for bookstore -> bookwarehouse
  - document mysql manifests
- other small tweaks and fixes

jaellio commented on content/docs/demos/ingress_k8s_nginx.md:
````diff
@@ -30,7 +30,7 @@ nginx_ingress_port="$(kubectl -n "$nginx_ingress_namespace" get service "$nginx_
 To restrict ingress traffic on backends to authorized clients, we will set up the IngressBackend configuration such that only ingress traffic from the endpoints of the Nginx Ingress Controller service can route traffic to the service backend. To be able to discover the endpoints of this service, we need OSM controller to monitor the corresponding namespace.
 ```bash
-kubectl label ns "$nginx_ingress_namespace" openservicemesh.io/monitored-by="$osm_mesh_name"
+osm namespace add "$nginx_ingress_namespace" --mesh-name "$osm_mesh_name" --osm-namespace "$osm_namespace" --disable-sidecar-injection
````
Is `--osm-namespace "$osm_namespace"` required for this command?

jaellio commented on content/docs/getting_started/manual_demo.md:
```diff
@@ -146,7 +147,7 @@ kubectl apply -f https://raw.githubusercontent.com/openservicemesh/osm/release-v
 Create the `bookstore` service account, service, and deployment:
```
`bookstore-v1`

jaellio replied:
bookstore-v1 instead of bookstore

jaellio commented on content/docs/getting_started/manual_demo.md:
```diff
@@ -81,31 +81,32 @@ Read more on OSM's integrations with Prometheus, Grafana, and Jaeger in the [obs
 
-The `bookbuyer`, `bookthief`, `bookstore`, and `bookwarehouse` Pods will be in separate Kubernetes Namespaces with
-the same names. Each new Pod in the service mesh will be injected with an Envoy sidecar container.
+The `bookbuyer`, `bookthief`, `bookstore-v1`, and `bookwarehouse` Pods will be in separate Kubernetes Namespaces. Each new Pod in the service mesh will be injected with an Envoy sidecar container.
```
We could note that the mysql pod will be deployed in the same namespace as bookwarehouse.

jaellio commented on content/docs/getting_started/manual_demo.md:
````diff
@@ -391,18 +395,25 @@ The counter in the `bookthief` window will stop incrementing.
 
 ```bash
 # Contains the bookstore-v2 Kubernetes Service, Service Account, Deployment and SMI Traffic Target resource to allow
 # bookbuyer to communicate with `bookstore-v2` pods
````
`bookbuyer` instead of bookbuyer

jaellio commented on content/docs/getting_started/traffic_policies.md:
```diff
@@ -0,0 +1,184 @@
+
+#### Allowing the Bookthief Application to access the Mesh
+
+Currently the Bookthief application has not been authorized to participate in the service mesh communication. We will now update the TrafficTarget to allow `bookthief` to communicate with `bookstore`.
```
bookstore-v1

jaellio commented on content/docs/getting_started/traffic_policies.md:
```diff
@@ -0,0 +1,184 @@
+spec:
+ destination:
+   kind: ServiceAccount
+   name: bookstore
```
bookstore-v1

jaellio commented on content/docs/getting_started/traffic_policies.md:
````diff
@@ -0,0 +1,184 @@
+kubectl apply -f https://raw.githubusercontent.com/openservicemesh/osm/release-v0.9/docs/example/manifests/access/traffic-access-v1.yaml
+```
+
+The counters should now be incrementing for the `bookbuyer`, and `bookstore` applications:
````
bookstore-v1

jaellio commented on content/docs/getting_started/traffic_split.md:
````diff
@@ -0,0 +1,80 @@
+To demonstrate usage of SMI traffic access and split policies, we will now deploy version v2 of the bookstore application (`bookstore-v2`) - remember that if you are using openshift, you must add the security context constraint to the bookstore-v2 service account as specified in the [installation guide](/docs/install/#openshift). A `bookstore` service will also be created to serve as the root of  ...
+
+```bash
+# Contains the bookstore-v2 Kubernetes Service, Service Account, Deployment and SMI Traffic Target resource to allow
````
The bookstore service is also contained in `bookstore-v2.yaml`. Could this be moved to a separate yaml?

jaellio commented on content/docs/getting_started/traffic_split.md:
```diff
@@ -0,0 +1,80 @@
+
+- [http://localhost:8082](http://localhost:8082) - **bookstore-v2**
+
+The counter should _not_ be incrementing because no traffic is flowing yet to the `bookstore-v2` service.
```
remove "yet". No traffic will ever flow to the `bookstore-v2` service, but rather the `bookstore` service.

shashankram commented on content/docs/getting_started/install_apps.md:
```diff
@@ -0,0 +1,144 @@
+
+- `bookbuyer` is an HTTP client making requests to `bookstore-v1`. This traffic is **permitted**.
+- `bookthief` is an HTTP client and much like `bookbuyer` also makes HTTP requests to `bookstore`. This traffic should be **blocked**.
+- `bookstore-v1` is a server, which responds to HTTP requests. It is also a client making requests to the `bookwarehouse` service.
```
It seems like the demo workflow is being updated. Previously the demo walked through permissive mode first which would deploy the `bookstore` app with the `bookstore-v1` identity, and then deploy bookstore-v2 to showcase traffic splitting with SMI. Isn't that the case anymore? Could you also describe at a high level in the commit messages what changes are being made to the demo?

nojnhuh replied:
I didn't intentionally make any significant changes to the overall steps of the demo. Since the old manual demo page is now split into multiple files, the order of the steps isn't super clear from just looking at the diffs, but the [preview](https://deploy-preview-195--osm-docs.netlify.app/docs/getting_started/) renders the pages in the right order.

The demo still walks through installing OSM in permissive mode, installs the apps and shows them running, then disables permissive mode and creates SMI policies.

shashankram replied:
It seems to me that there's a difference. Previously, the demo starts with permissive mode which deploys the `bookstore` application where `bookstore.yaml` is used (not `bookstore-v1`):
![image](https://user-images.githubusercontent.com/21697719/135498674-aa0cd438-50b5-4c3c-9934-b2dce51f6728.png)

With your changes, the app has been changed to `bookstore-v1` where `bookstore-v1.yaml` is used:
![image](https://user-images.githubusercontent.com/21697719/135498775-806feb25-14a7-43a1-b184-494ae396843f.png)

I am unsure if this was intentional, but there is a reason why permissive mode used `bookstore` and not `bookstore-v1`. It's fine if we want to change that, but I'd like to see that surfaced in the commit message and the reason for doing so.

nojnhuh replied:
I essentially renamed the old bookstore.yaml to bookstore-v1.yaml, then added a new bookstore.yaml that's just the root bookstore service. The most significant change here is aligning the manifests closer to the automated demo, which has a bookstore root service and bookstore-v1 and bookstore-v2 backend services. To compensate for this change, I've updated the bookbuyer (and bookthief) to [target the bookstore-v1 service](https://github.com/openservicemesh/osm/pull/4180/files#diff-482c2abe2dc68b82c84b7fc63d58804898a70d49a2ac1f3b503ef08b0a48588eR41) initially, then [update that later](https://github.com/openservicemesh/osm-docs/pull/195/files#diff-91a064e053df3695c52c91b719f85244b867e4f3e3f13ff3a0d6847c11f68e48R24-R29) to point to the bookstore root service once that's been created. Does this address what you mean by "there is a reason why permissive mode used `bookstore` and not `bookstore-v1`"?

shashankram replied:
The goal should not be to align the manual demo to the automated demo, because the automated demo is not user-friendly. In the past, I had worked to address this issue by reworking the manual demo so that the onboarding experience for users is clear, which is not the case with the automated demo (see https://github.com/openservicemesh/osm/commit/54b7749edd7e3b12ecfca403805e1b581c4c8a6a).

The [update that later section](https://github.com/openservicemesh/osm-docs/pull/195/files#diff-91a064e053df3695c52c91b719f85244b867e4f3e3f13ff3a0d6847c11f68e48R24-R29) does not make sense to me. Users aren't going to be able to update their application binary in a way that our demo manifests allow. We need to be able to demonstrate the following in this order:
1. Install OSM
2. Enable permissive traffic policy mode
3. Deploy apps
4. Demonstrate that bookbuyer and bookthief apps can access the `bookstore-v1` service or `bookstore` service. If the `bookstore` service is used, then there must be a pod backing the `bookstore` service for permissive mode to function.
5. Switch to SMI mode and verify that all traffic is blocked
6. Deploy TrafficTarget to allow bookbuyer to access bookstore, and showcase that bookthief cannot access bookstore
7. Document TrafficSplit usage. The recommended way should be to deploy a second version of the app while the first app is still running. If this isn't possible, then it must be called out. Requiring users to update thier app binaries to use traffic split is a serious caveat, so if we can demo this in a way where it's avoidable, we should explore that.

nojnhuh replied:
> Users aren't going to be able to update their application binary in a way that our demo manifests allow.

This is true, but the steps I added only involve changing an environment variable on the bookbuyer and bookthief containers.

These changes still adhere to the steps you describe AFAICT. Where do you see that they don't?

shashankram replied:
I did not suggest that the doc steps don't adhere to the steps described above, but renaming bookstore -> bookstore-v1 in the docs and environment variables seems confusing enough to the extent that I suggest making that change in a separate PR. I'd really like for this PR to only refactor the getting-started page, not the demo docs. If you want to make changes to the demo docs, let's carve it out and make it a separate change so we can clearly demonstrate caveats with respect to changing the request URLs. Some of the problems I am highlighting might be present in the existing version of the docs as well, which I think must be clearly documented for v1.0.

nojnhuh replied:
We've decided not to go forward with the bookstore->bookstore-v1 changes. Those specific changes have been removed.

shashankram reviewed (dismissed):
The documentation changes in the manual demo aren't clear to me. Discussion thread: https://github.com/openservicemesh/osm-docs/pull/195#discussion_r719596853

nojnhuh commented:
Converting this to a draft now that this is dependent on #202 which is a subset of these changes.

nojnhuh commented:
Since we've decided not to go forward with the changes in #202, I've updated this PR to remove those, and also split up the changes into more commits. I've updated the PR description to capture these changes.

jaellio commented on content/docs/getting_started/cleanup.md:
````diff
@@ -0,0 +1,22 @@
+
+To uninstall OSM, run
+```bash
+osm uninstall
````
To avoid the cli prompt to confirm which mesh to uninstall we could specify `--mesh-name osm`.

shashankram replied:
We don't want to assume what the mesh-name is. To avoid the prompt, use `osm uninstall -f`.

nojnhuh replied:
I think letting users work through the prompt might actually be preferable here to show that it exists.

jaellio replied:
I had assumed the mesh name was `osm` since in the install guide `content/docs/getting_started/install_osm.md` we use the default mesh name. If we are not going to make this assumption then I think it makes sense for the users to work through the prompt.

phillipgibson commented:
@nojnhuh I would suggest renaming "Install OSM" under the getting started to "Setup OSM". This install can be confused for the Installation under the Guides section.

michelleN commented on content/docs/demos/outbound_ip_exclusion.md:
```diff
@@ -35,7 +35,7 @@ The following demo shows an HTTP `curl` client making HTTP requests to the `http
 
     # Deploy curl client in the curl namespace
-    kubectl apply -f docs/example/manifests/samples/curl/curl.yaml -n curl
+    kubectl apply -f https://raw.githubusercontent.com/openservicemesh/osm/main/docs/example/manifests/samples/curl/curl.yaml -n curl
```
Should this url point to the yaml file on a release branch instead of on main?

nojnhuh replied:
Based on #191 we decided that docs on the main branch here link to the main branch of the osm repo.

michelleN commented on content/docs/getting_started/install_apps.md:
````diff
@@ -0,0 +1,145 @@
+Create the `mysql` service account, service, and stateful set:
+
+```bash
+kubectl apply -f https://raw.githubusercontent.com/openservicemesh/osm/release-v0.9/docs/example/manifests/apps/mysql.yaml
````
this link didn't work for me

michelleN commented on content/docs/getting_started/install_apps.md:
````diff
@@ -0,0 +1,145 @@
+To view these resources on your cluster, run the following commands:
+
+```bash
+kubectl get deployments -n bookbuyer
````
kubectl get pods,deployments,statefulset,services,endpoints --namespace bookbuyer,boothief,bookwarehouse,bookstore

michelleN commented on content/docs/guides/install.md:
````diff
@@ -102,7 +102,7 @@ $ kubectl get pods,svc,secrets,meshconfigs,serviceaccount --namespace osm-system
 
 ```console
-kubectl get clusterrolebinding,clusterrole,mutatingwebhookconfiguration
+kubectl get clusterrolebinding,clusterrole,mutatingwebhookconfiguration,validatingwebhookconfigurations -l app.kubernetes.io/name=openservicemesh.io
````
o0o nice

michelleN commented on content/docs/getting_started/install_apps.md:
````diff
@@ -0,0 +1,145 @@
+
+```bash
+osm namespace add bookstore
+osm namespace add bookbuyer
````
`osm namespace add bookstore bookbuyer bookthief bookwarehouse`

michelleN commented on content/docs/getting_started/install_apps.md:
```diff
@@ -0,0 +1,145 @@
+kubectl get services -n bookwarehouse
+
+kubectl get endpoints -n bookstore
+kubectl get endpoints -n bookwarehouse
```
```bash
kubectl get pods,deployments,serviceaccounts --n bookbuyer
kubectl get pods,deployments,serviceaccounts --n bookthief

kubectl get pods,deployments,serviceaccounts,services,endpoints --n bookstore
kubectl get pods,deployments,serviceaccounts,services,endpoints --n bookwarehouse

nojnhuh commented on content/docs/demos/ingress_contour.md:
````diff
@@ -33,12 +33,15 @@ osm install --set contour.enabled=true \
 To restrict ingress traffic on backends to authorized clients, we will set up the IngressBackend configuration such that only ingress traffic from the endpoints of the `osm-contour-envoy` service can route traffic to the service backend. To be able to discover the endpoints of `osm-contour-envoy` service, we need OSM controller to monitor the corresponding namespace.
 ```bash
-kubectl label ns "$osm_namespace" openservicemesh.io/monitored-by="$osm_mesh_name"
+osm namespace add "$osm_namespace" --mesh-name "$osm_mesh_name" --disable-sidecar-injection
````
The short answer is to preserve the behavior of the `kubectl` command this replaces. I would need help coming up with a better reasoning than "otherwise Contour won't work." Is something like that all you would expect to see here?

nojnhuh commented on content/docs/getting_started/install_apps.md:
```diff
@@ -0,0 +1,145 @@
+
+- `bookbuyer` is an HTTP client making requests to `bookstore`. This traffic is **permitted**.
+- `bookthief` is an HTTP client and much like `bookbuyer` also makes HTTP requests to `bookstore`. This traffic should be **blocked**.
+- `bookstore` is a server, which responds to HTTP requests. It is also a client making requests to the `bookwarehouse` service.
```
Done

nojnhuh commented on content/docs/getting_started/install_apps.md:
```diff
@@ -0,0 +1,145 @@
+- `mysql` is a MySQL database only reachable by `bookwarehouse`.
+
+
+We are going to craft SMI policies, which will bring us to this final desired
```
Done

nojnhuh commented on content/docs/getting_started/install_apps.md:
```diff
@@ -0,0 +1,145 @@
+| mysql         |     ❌    |     ❌    |     ❌    |      ❌       |  n/a  |
+
+
+To show SMI Traffic Split, we will deploy an additional application:
```
Done
